### DISCUSSION TITLE
How to get both `Production` and `parse_tree`?

### DISCUSSION THREAD

**Discussion by Life4gal** [foonathan/lexy#99] [Q&A] [answered]
We can print the result of source file parsing through [visualize](https://lexy.foonathan.net/reference/visualize/#visualize).
But the printed object must be [parse_tree](https://lexy.foonathan.net/reference/parse_tree/#parse_tree), which needs to be generated by [parse_as_tree](https://lexy.foonathan.net/reference/action/parse_as_tree/#parse_as_tree).

But it will discard all generated values.
![image](https://user-images.githubusercontent.com/52756109/203489998-837dd4a3-1a75-4e7a-b134-a5a1479c4281.png)

To get the result of the parsing, we need [parse](https://lexy.foonathan.net/reference/action/parse/#parse). (I tried to store the results in the state, so that the return of `parse` would be insignificant. But I found that the state also had no values.)

This causes that if you want to get both the result and parse_tree, you must parse them twice.
```c++
auto file = lexy::read_file<lexy::utf8_encoding>(filename.data());

if (!file)
{
	throw std::runtime_error{"Cannot read file!"};
}

ParseState state{string::string{filename}, std::move(file).buffer()};

if (const auto result =
			lexy::parse<grammar::module_declaration>(state.buffer, state, lexy_ext::report_error.opts({.flags = lexy::visualize_fancy}).path(state.filename.c_str()));
	!result.is_success())
{
	throw std::runtime_error{"Cannot parse file!"};
}

#ifdef PARSE_AS_TREE
lexy::parse_tree_for<decltype(state.buffer)> tree{};
lexy::parse_as_tree<grammar::module_declaration>(tree, state.buffer, state, lexy_ext::report_error.opts({.flags = lexy::visualize_fancy}).path(state.filename.c_str()));
lexy::visualize(stdout, tree, {lexy::visualize_fancy});
#endif
```

**Comment by foonathan** [comment]
That's not supported by the library at the moment. I could add a `lexy::parse_with_tree` function that does it, but want to understand the use case further.

Why do you need both the tree and the value? If you have the tree, you have all the information you need to compute the values yourself.

**Reply by Life4gal** [reply]
In fact, there are no use cases that use this feature, but if there is such a function, it is much easier to write code in debugging mode. After all, you can clearly see the results of source file parsing. 😄

**Reply by foonathan** [reply]
Ah, if it's for debugging, I might rather add the ability to combine tracing with parsing: https://lexy.foonathan.net/reference/action/trace/#trace

Would that work for you as well?

**Reply by Life4gal** [reply]
I'm not sure what I missed, I don't quite understand what you mean.

Am I supposed to be like this?
```c++
auto file = lexy::read_file<lexy::utf8_encoding>(filename.data());

if (!file)
{
	throw std::runtime_error{"Cannot read file!"};
}

ParseState state{string::string{filename}, std::move(file).buffer()};

#ifdef TRACE_PARSE
lexy::trace<grammar::module_declaration>(
				stdout,
				state.buffer,
				state,
				{.flags = lexy::visualize_fancy});
#else
if (const auto result =
			lexy::parse<grammar::module_declaration>(state.buffer, state, lexy_ext::report_error.opts({.flags = lexy::visualize_fancy}).path(state.filename.c_str()));
	!result.is_success())
{
	throw std::runtime_error{"Cannot parse file!"};
}
#endif
```
Or this?
```c++
auto file = lexy::read_file<lexy::utf8_encoding>(filename.data());

if (!file)
{
	throw std::runtime_error{"Cannot read file!"};
}

ParseState state{string::string{filename}, std::move(file).buffer()};

if (const auto result =
			lexy::parse<grammar::module_declaration>(state.buffer, state, lexy_ext::report_error.opts({.flags = lexy::visualize_fancy}).path(state.filename.c_str()));
	!result.is_success())
{
	throw std::runtime_error{"Cannot parse file!"};
}

#ifdef TRACE_PARSE
lexy::trace<grammar::module_declaration>(
				stdout,
				state.buffer,
				state,
				{.flags = lexy::visualize_fancy});
#endif
```

I noticed that `trace` does not produce any value, either the `return value` or the `value of state`. 
So obviously the first piece of code is wrong, but if it is the second piece of code, what is the difference between it and `parse_as_tree`?

**Reply by Life4gal** [reply]
My idea is whether there is a way to get both the `parsed value` and `print the parse tree` at the `same time` during a source file parsing.

**Reply by foonathan** [reply]
Yes, currently that's not supported.

My question was: do you explicitly want to dump the parse tree, or is tracing okay as well? Since tracing does not return a value currently it's easier to compose with an existing parse action.

The idea is to enable an interface like `lexy::trace<grammar::module_declaration>(lexy::parse_action(state, report_error), stdout, input)` which traces parsing and returns the value.

**Reply by Life4gal** [reply]
Tracing is also OK. In fact, I just hope that there is a way to print the parse tree (if necessary) at the same time when get the value of the parse source file.

**Comment by foonathan** [comment, marked as answer]
I've started implementing something like that, but it's not trivial. I don't think it's necessary: if you just want to visualize the parse tree for debugging purposes, just parse it twice - performance doesn't matter there anyway.
